packages/contracts-bridge/contracts/test/TokenRegistry.t.sol

```diff
@@ -1,10 +1,558 @@
+            tokenRegistry.ensureLocalToken(newDomain, newId),
+            customAddress
+        );
+        vm.stopPrank();
+    }
+
+    function test_enrollCustomOnlyOwner() public {
+        uint32 newDomain = 24;
+        bytes32 newId = "yaw";
+        address customAddress = address(0xBEEF);
+        vm.expectRevert("Ownable: caller is not the owner");
+        tokenRegistry.enrollCustom(newDomain, newId, customAddress);
+    }
+
+    function test_enrollCustomFuzzed(
+        uint32 dom,
+        bytes32 id,
+        address addr
+    ) public {
+        vm.prank(tokenRegistry.owner());
+        tokenRegistry.enrollCustom(dom, id, addr);
+        (uint32 storedDomain, bytes32 storedId) = tokenRegistry
+            .getCanonicalTokenId(addr);
+        address storedAddress = tokenRegistry.getRepresentationAddress(dom, id);
+        assertEq(uint256(storedDomain), uint256(dom));
+        assertEq(storedId, id);
+        assertEq(storedAddress, addr);
+    }
+
+    function test_oldReprToCurrentRepr() public {
+        uint32 domain = 24;
+        bytes32 id = "yaw";
+        address oldAddress = address(0xBEEF);
+        vm.prank(tokenRegistry.owner());
+        // Enroll first local custom token of a remote asset
+        tokenRegistry.enrollCustom(domain, id, oldAddress);
+        address newAddress = address(0xBEEFBEEF);
+        vm.prank(tokenRegistry.owner());
+        // After some time, the owner wants to enroll a  new implementation of the local custom token
+        // for the same remote asset
+        tokenRegistry.enrollCustom(domain, id, newAddress);
+        // We make sure that a user can retrieve the new implementation of the remote asset, using the old
+        // implementation as a key
+        assertEq(tokenRegistry.oldReprToCurrentRepr(oldAddress), newAddress);
+    }
+
+    function test_oldReprToCurrentReprFuzzed(
+        address oldAddress,
+        address newAddress,
+        uint32 domain,
+        bytes32 id
+    ) public {
+        vm.assume(domain != 0);
+        vm.prank(tokenRegistry.owner());
+        tokenRegistry.enrollCustom(domain, id, oldAddress);
+        vm.prank(tokenRegistry.owner());
+        tokenRegistry.enrollCustom(domain, id, newAddress);
+        assertEq(tokenRegistry.oldReprToCurrentRepr(oldAddress), newAddress);
+    }
+
+    function test_getTokenIdCanonical() public {
+        (uint32 domain, bytes32 id) = tokenRegistry.getTokenId(
+            address(localToken)
+        );
+        assertEq(uint256(domain), localDomain);
+        assertEq(id, TypeCasts.addressToBytes32(address(localToken)));
+    }
+
+    function test_getTokenIdRepr() public {
+        (uint32 domain, bytes32 id) = tokenRegistry.getTokenId(
+            address(remoteTokenLocalAddress)
+        );
+        assertEq(uint256(domain), remoteDomain);
+        assertEq(id, remoteTokenRemoteAddress);
+    }
+
+    function test_getTokenIdRerprFuzzed(uint32 dom, bytes32 id) public {
+        // Domain can be 0?
```

Comment:
yes in theory. not in practice. just assume not 0

Reply:
I would say not in theory as well, multiple contracts break with domain = 0. 

I think it's an invariant of the system